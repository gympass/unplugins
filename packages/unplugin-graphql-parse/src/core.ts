import { parse } from 'graphql';

import type { Options } from './options';

export function transform(code: string) {
  const schema = JSON.stringify(parse(code));

  const contents = `const schema = ${schema};
export default schema;
`;

  return contents;
}

export function transformInclude(id: string, options: Options) {
  return id.endsWith(options.ext);
}
