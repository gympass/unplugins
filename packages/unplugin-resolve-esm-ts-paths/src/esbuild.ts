import unplugin from './index.js';

export default unplugin.esbuild;
