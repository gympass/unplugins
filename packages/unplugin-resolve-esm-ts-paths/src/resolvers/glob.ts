import type { ImportSpecifier } from 'es-module-lexer';
import { relative } from 'node:path';
import glob from 'tiny-glob';
import { removeExt } from '../helpers/index.js';

export async function resolveGlob(
  id: string,
  fileImport: ImportSpecifier,
  dir: string,
  endExt: string,
  importString: string,
) {
  const file = fileImport.n as string;

  const files = await glob(file.split(':')[1], {
    cwd: dir,
    filesOnly: true,
  });

  const selfPath = relative(dir, id);
  const filteredFiles = files.filter(f => f !== selfPath);

  const modulesImports = filteredFiles
    .map(
      (module, index) =>
        // Since this is the transpiled code, we can't have
        // extensions other than .js. That's why we replace everything.
        `import module${index} from './${removeExt(module)}${endExt}'`,
    )
    .join('\n');

  const modules = filteredFiles.map((_, index) => `module${index}`).join(', ');

  const variable = (
    importString.match(/import (.*) from/) as RegExpMatchArray
  ).pop();

  const newCode = `
    ${modulesImports}
    const ${variable} = [${modules}]
  `;

  return [importString, newCode] as [string, string];
}
