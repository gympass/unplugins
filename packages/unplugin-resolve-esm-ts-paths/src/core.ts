import { parse } from 'es-module-lexer';

import { dirname } from 'node:path';

import type { Options } from './options.js';
import { resolveGlob } from './resolvers/glob.js';
import { resolveRelative } from './resolvers/relative.js';
import { resolvePaths } from './resolvers/paths.js';
import {
  BROKEN_IMPORT,
  MODULE_IMPORT,
  escapePathForRegExp,
} from './helpers/index.js';

export async function transform(code: string, id: string, options: Options) {
  const dir = dirname(id);

  const endExt = options.resolvedExt;

  try {
    const [imports] = await parse(code);

    if (imports.length === 0) return code;

    const replaceMapPromises = imports.map<Promise<[string, string]>>(
      fileImport => {
        const file = fileImport.n;
        if (!file) return Promise.resolve([BROKEN_IMPORT, BROKEN_IMPORT]);

        const importString = code.substring(fileImport.ss, fileImport.se);

        const correctQuote = (
          /'|"/.exec(importString) as RegExpExecArray
        ).pop() as string;

        // Resolve Glob imports.
        if (options.resolveGlob && file.includes('glob:')) {
          return resolveGlob(id, fileImport, dir, endExt, importString);
        }

        // Add index file import for relative directory imports.
        if (file.includes('./')) {
          return resolveRelative(file, dir, endExt, correctQuote);
        }

        return resolvePaths(file, dir, endExt, correctQuote);
      },
    );

    // Clear the map keeping unique keys and remove module imports.
    const replaceMap = new Map<string, string>(
      await Promise.all(replaceMapPromises),
    );
    replaceMap.delete(MODULE_IMPORT);
    replaceMap.delete(BROKEN_IMPORT);

    // If size is zero, there are only modules imports.
    if (replaceMap.size === 0) return code;

    const allPathsRegExp = new RegExp(
      [...replaceMap.keys()].map(escapePathForRegExp).join('|'),
      'g',
    );

    return code.replace(allPathsRegExp, match => {
      return replaceMap.get(match) as string;
    });
  } catch (error) {
    /**
     * If we get a problem during parsing, we return the original
     * code so the bundler can handle it.
     */
    return code;
  }
}

export function transformInclude(id: string, options: Options) {
  if (options.ext instanceof RegExp) {
    return Boolean(id.match(options.ext));
  }
  return id.endsWith(options.ext);
}
