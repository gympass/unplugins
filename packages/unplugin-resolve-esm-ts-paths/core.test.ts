import { transform, transformInclude } from './src/core.js';
import type { Options } from './src/options.js';

const code = `
  import foo from './foo';
  import doubleQ from "./doubleQ";
  import { build } from 'esbuild'
  import item from '@test/file'
  import glob from 'glob:./*.md'
`;

it('should accept the incomming ext', () => {
  expect(transformInclude('foo.ts', { ext: '.ts' } as Options)).toBeTruthy();
  expect(transformInclude('foo.ts', { ext: '.js' } as Options)).toBeFalsy();
});

it('should resolve imports', async () => {
  const r = await transform(code, 'foo.ts', {
    ext: '.ts',
    resolveGlob: true,
    resolvedExt: '.js',
  });

  expect(r).toMatchSnapshot();
});
