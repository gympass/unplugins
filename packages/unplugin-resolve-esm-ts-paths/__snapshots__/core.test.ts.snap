// Vitest Snapshot v1

exports[`should resolve imports 1`] = `
"
  import foo from './foo.js';
  import doubleQ from './doubleQ.js';
  import { build } from 'esbuild'
  import item from './src/test/file.js'
  
    import module0 from './README.js'
    const glob = [module0]
  
"
`;
